src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -134,9 +138,13 @@ public JCTree.JCMethodDecl transformFuncOpToAST(CoreOp.FuncOp funcOp, Name metho
         for (Value root : rootValues) {
             JCTree tree = opToTree(root);
             if (tree instanceof JCExpression e) {
-                var vs = new Symbol.VarSymbol(LocalVarFlags | SYNTHETIC, names.fromString("_$" + localVarCount++), tree.type, ms);
-                tree = treeMaker.VarDef(vs, e);
-                valueToTree.put(root, tree);
+                if (!root.uses().isEmpty()) {
+                    var vs = new Symbol.VarSymbol(LocalVarFlags | SYNTHETIC, names.fromString("_$" + localVarCount++), tree.type, ms);
+                    tree = treeMaker.VarDef(vs, e);
+                    valueToTree.put(root, tree);
+                } else {
+                    tree = treeMaker.Exec(e);
```

Comment:
This is the heart of the fix - if there's no usage, just emit as expression statement. I have tried with an alternate fix which, in case of exactly one use, would "inline" the use (avoiding the local). While this delivered massive reduction in the number of locals, the generated model was no longer compatible as some values were being reordered.